src/Mono.Android/Android.Runtime/AndroidRuntime.cs

```diff
@@ -530,20 +558,78 @@ public void RegisterNativeMembers (JniType nativeClass, Type type, ReadOnlySpan<
 							while (callbackDeclaringType.ContainsGenericParameters) {
 								callbackDeclaringType = callbackDeclaringType.BaseType!;
 							}
-							GetCallbackHandler connector = (GetCallbackHandler) Delegate.CreateDelegate (typeof (GetCallbackHandler),
-								callbackDeclaringType, callbackString.ToString ());
-							callback = connector ();
+#if ENABLE_MARSHAL_METHODS
+							// TODO: this is temporary hack, it needs a full fledged registration mechanism for methods like these (that is, ones which
+							// aren't registered with [Register] but are baked into Mono.Android's managed and Java code)
+							bool createCallback = false;
+							string declaringTypeName = callbackDeclaringType.FullName;
+							string callbackName = callbackString.ToString ();
+
+							foreach (var kvp in dynamicRegistrationMethods) {
+								string dynamicTypeName = kvp.Key;
+
+								foreach (string dynamicCallbackMethodName in kvp.Value) {
+									if (ShouldRegisterDynamically (declaringTypeName, callbackName, dynamicTypeName, dynamicCallbackMethodName)) {
+										createCallback = true;
+										break;
+									}
+								}
+
+								if (createCallback) {
+									break;
+								}
+							}
+
+							if (createCallback) {
+								Logger.Log (LogLevel.Info, "monodroid-mm", $"  creating delegate for: '{callbackString.ToString()}' in type {callbackDeclaringType.FullName}");
+#endif
+								GetCallbackHandler connector = (GetCallbackHandler) Delegate.CreateDelegate (typeof (GetCallbackHandler),
+								                                                                             callbackDeclaringType, callbackString.ToString ());
+								callback = connector ();
+#if ENABLE_MARSHAL_METHODS
+							} else {
+								Logger.Log (LogLevel.Warn, "monodroid-mm", $"  would try to create delegate for: '{callbackString.ToString()}' in type {callbackDeclaringType.FullName}");
+							}
+#endif
+						}
+
+						if (callback != null) {
+#if ENABLE_MARSHAL_METHODS
+							needToRegisterNatives = true;
+#endif
+							natives [nativesIndex++] = new JniNativeMethodRegistration (name.ToString (), signature.ToString (), callback);
 						}
-						natives [nativesIndex++] = new JniNativeMethodRegistration (name.ToString (), signature.ToString (), callback);
 					}
 
 					methodsSpan = newLineIndex != -1 ? methodsSpan.Slice (newLineIndex + 1) : default;
 				}
 
-				JniEnvironment.Types.RegisterNatives (nativeClass.PeerReference, natives, natives.Length);
+#if ENABLE_MARSHAL_METHODS
+				if (needToRegisterNatives) {
+					// We need to reallocate as JniEnvironment.Types.RegisterNatives uses a `foreach` loop and will NREX otherwise (since we aren't filling all
```

Comment:
We should fix Java.Interop so that null `.Marshaler` values are either allowed, or allowed after `numMethods`.

Rephrased: [`JniEnvironment.Types.RegisterNatives()`](https://github.com/xamarin/java.interop/blob/6d1ae4ee1076341f8af36dbdd306dce0e6266999/src/Java.Interop/Java.Interop/JniEnvironment.Types.cs#L204-L211) should use a `for` loop instead:

```csharp
for (int i = 0; i < numMethods; ++i) {
    var m = methods [i];
    …
}
```

Reply:
Should be fixed by: https://github.com/xamarin/java.interop/pull/1029